walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java

```diff
@@ -14,10 +14,72 @@
 
 package com.google.sps;
 
+import java.util.ArrayList;
 import java.util.Collection;
+import java.util.Collections;
+import java.util.LinkedList;
 
 public final class FindMeetingQuery {
+
   public Collection<TimeRange> query(Collection<Event> events, MeetingRequest request) {
-    throw new UnsupportedOperationException("TODO: Implement this method.");
+
+    // Using linked list as removing the first element will take O(1) time
+    LinkedList<Event> eventList = new LinkedList<Event>(events);
+    Collections.sort(eventList, Event.ORDER_EVENT_BY_START); // Collections.sort() -> O(nlogn)
+
+    ArrayList<TimeRange> s = new ArrayList<TimeRange>();
+
+    long currentTime = (long) TimeRange.START_OF_DAY;
+
+    // making sure the request is within bounds
+    if (request.getDuration() > TimeRange.WHOLE_DAY.duration()) {
+      return s;
+    }
+
+    if (request.getDuration() < 0) {
+      return s;
+    }
+
+    while (currentTime < (long) TimeRange.END_OF_DAY) {
+
+      if (eventList.isEmpty()) {
+        TimeRange range = TimeRange.fromStartEnd((int) currentTime, TimeRange.END_OF_DAY, true);
+        s.add(range);
+        break;
+      }
+
+      // if there are no common people between the request and the event
+      if (Collections.disjoint(eventList.getFirst().getAttendees(), request.getAttendees())) {
+        eventList.removeFirst();
+        continue;
+      }
+      if ((currentTime + request.getDuration()) <= eventList.getFirst().getStartTime()) {
+        TimeRange range =
+            TimeRange.fromStartEnd(
+                (int) currentTime, (int) eventList.getFirst().getStartTime(), false);
+        s.add(range);
+        currentTime = eventList.getFirst().getStartTime();
```

Comment:
My entire algorithm works with the top of the "stack" which is really a linked list so that I can retrieve values in O(1) time. I'd prefer to keep this bc I understand it more, but if you think it's better to change it I will make it work

Reply:
Ah, after rereading this to clarify, I actually do not remove this from the list due to the case where Person A has a meeting, but person B and C need to meet. In my linked list, the events are not distinguished by person. So in the case where Person A has a meeting, but person B and C need to meet, the currentTime will at the start of that meeting, but when the loop iterates again, that element will be removed since there are no matching attendees